src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,544 @@
+                    {
+                        break;
+                    }
+                    index = entry.Next;
+                }
+            }
+            return index;
+        }
+
+        private bool TryInsert(int? index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            int i = IndexOf(key, out int hashCode);
+            Entry[] entries = _entries;
+            if (i >= 0)
+            {
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        entries[i].Value = value;
+                        return true;
+                    case InsertionBehavior.ThrowOnExisting:
+                        throw new ArgumentException(string.Format(Strings.Argument_AddingDuplicateWithKey, key.ToString()));
+                    default:
+                        return false;
+                }
+            }
+
+            ++_version;
+            // Check if resize is needed
+            int count = Count;
+            if (entries.Length == count)
+            {
+                Resize(HashHelpers.ExpandPrime(entries.Length));
+                entries = _entries;
+            }
+
+            // Increment indices >= index;
+            int actualIndex = index ?? count;
+            for (i = actualIndex; i < count; ++i)
+            {
+                UpdateBucketIndices(i, 1);
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            ref int bucket = ref _buckets[hashCode % _buckets.Length];
+            Entry newEntry = new Entry { HashCode = hashCode, Key = key, Value = value, Next = bucket - 1 };
+            entries[actualIndex] = newEntry;
+            bucket = actualIndex + 1;
+            ++Count;
+            return true;
+        }
+
+        private void RemoveEntryFromBucket(int i)
+        {
+            Entry[] entries = _entries;
+            Entry entry = entries[i];
+            ref int b = ref _buckets[entry.HashCode % _buckets.Length];
+            if (b == i + 1)
+            {
+                b = entry.Next + 1;
+            }
+            else
+            {
+                int j = b - 1;
+                do
+                {
+                    ref Entry e = ref entries[j];
+                    if (e.Next == i)
+                    {
+                        e.Next = entry.Next;
+                        j = -1;
+                    }
+                    else
+                    {
+                        j = e.Next;
+                    }
+                } while (j >= 0);
+            }
+        }
+
+        private void UpdateBucketIndices(int i, int incrementAmount)
+        {
+            Entry[] entries = _entries;
+            Entry entry = entries[i];
+            ref int b = ref _buckets[entry.HashCode % _buckets.Length];
+            if (b == i + 1)
+            {
+                b += incrementAmount;
```

Comment:
Simliar comments to above?

Thanks for breaking the method into two, now it's clear to me what it does. Maybe there is a way to write it in one method in a clear enough way, but this is fine for now.